### PR TITLE
common/admin_socket: Admin socket self reflection

### PR DESCRIPTION
**Pull request by aclamk**
This allows for ceph programs and utilities to use admin_socket inspection engine to provide self-reflection on current state. 
This can be used to any purpose, but this is devised to be used by ceph-objectstore-tool. Ceph-objectstore-tool will be able to query fragmentation without necessity of making extra interfaces in Allocator class.

### CODE REVIEW

File: src/common/admin_socket.cc [hunk 1]
```diff
@@ -336,23 +336,42 @@ bool AdminSocket::do_accept()
...
+      if (out.write_fd(connection_fd) >= 0)
+        rval = true;
+    }
+  }
 
+  retry_sys_call(::close, connection_fd);
+  return rval;
+}
+
+bool AdminSocket::execute_command(const std::string& cmd, ceph::bufferlist& out)
```
Comment (tchaikov) [inline src/common/admin_socket.cc]: nit, could return a `pair<bool, out>` and use structured binding to capture the return values.

also, the commit message could be more specific -- s/Add/Extract/

File: src/common/admin_socket.cc [hunk 2]
```diff
@@ -377,53 +395,41 @@ bool AdminSocket::do_accept()
...
-    uint32_t len = htonl(out.length());
-    int ret = safe_write(connection_fd, &len, sizeof(len));
-    if (ret < 0) {
-      lderr(m_cct) << "AdminSocket: error writing response length "
-		   << cpp_strerror(ret) << dendl;
-    } else {
-      if (out.write_fd(connection_fd) >= 0)
-	rval = true;
-    }
+    lderr(m_cct) << "AdminSocket: request '" << cmd << "' not defined" << dendl;
```
Comment (tchaikov) [inline src/common/admin_socket.cc]: nit, could take this chance to put:
```c++
lderr(m_cct) << "AdminSocket: request " << quoted(cmd) << "not defined" << dendl;
```

File: src/common/admin_socket.cc [hunk 3]
```diff
@@ -377,53 +395,41 @@ bool AdminSocket::do_accept()
...
+    out.append("failed");
+  } else {
+    ldout(m_cct, 5) << "AdminSocket: request '" << match << "' '" << args
+        << "' to " << match_hook
+        << " returned " << out.length() << " bytes" << dendl;
+  }
+  return true;
 }
 
+
```
Comment (tchaikov) [inline src/common/admin_socket.cc]: please drop these two new empty lines.

**Comment by jdurgin** [comment]
This was merged as part of #29425

**Pull request closed without merging**